fix: Extra column in csv(or non-raw csv) get while sending on email

rohitpawar2811 opened the pull request:
### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
While sending csv on Email getting an extra column(non-raw csv get) which has a count from 0 to n-row .
Which I should not want.( And I think no one wants in CSV because we get numbering while opening it on platforms )

I have fix this thing and added CSV_INDEX=False //In superset_config.py
Which make to not create an index 
If You want index in csv make it True.

For More detail on this fix Refer this issue
https://github.com/apache/superset/issues/22981


### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

- Simply Checked By sending chart as csv on email extra index-column is not coming

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: Fixes 22981
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

rohitpawar2811 committed:
fix : We are geeting an extra column in csv while sending csv on email which is caused by post_processing.py and now it is Fixed (\#22981)

rohitpawar2811 committed:
Update config.py

Corrected my message

Antonio-RiveroMartnez reviewed (commented):
Not sure if possible, but could we add some tests for the different values (true/false) and their outputs?

betodealmeida reviewed (changes requested):
This looks great! I just left a few comments.

Rohit-pawar902 commented on superset/config.py:
```diff
@@ -748,6 +748,10 @@ def _try_json_readsha(filepath: str, length: int) -> Optional[str]:
 # note: index option should not be overridden
 CSV_EXPORT = {"encoding": "utf-8"}
 
+#It Is used for enable/disable the index sequence for csv which came at email-time,
+#rearly someone want the index in csv so by-default it is false
+CSV_INDEX={"index":False}
+
```
@betodealmeida Your all changes looks good and thanks for responce. 
But According to this comment :
https://github.com/apache/superset/issues/22981#issuecomment-1440410419
I would thing we don't have to give config `opt true/false` , instead we can make this  `processed_df.to_csv(buf, index=config["CSV_INDEX"]) ` to processed_df.to_csv(buf, index=False) Because it is really a bug.
what do you think ? I like to listen your thougts regarding to this.
@Antonio-RiveroMartnez

Antonio-RiveroMartnez replied:
Hey @rohitpawar2811 ! we could use `True/False` in case some users got used to the `row_number` already? as @betodealmeida said, we can change that in future releases along addressing the other problems listed in the issue. Let's add some tests for both cases tho. Thanks again!

rohitpawar2811 committed:
Update superset/charts/post_processing.py

Co-authored-by: Beto Dealmeida <roberto@dealmeida.net>

rohitpawar2811 committed:
Update superset/charts/post_processing.py

Co-authored-by: Beto Dealmeida <roberto@dealmeida.net>

rohitpawar2811 committed:
Update superset/config.py

Co-authored-by: Beto Dealmeida <roberto@dealmeida.net>

rohitpawar2811 commented:
`Hey @rohitpawar2811 ! we could use True/False in case some users got used to the row_number already? as @betodealmeida said, we can change that in future releases along addressing the other problems listed in the issue. Let's add some tests for both cases if possible. Thanks again!`
got your point related True/False
Now  For Test-Case :
1. I think we can create chart with sample data and then  
2. we have to simple call the api(/api/v1/chart/628/data/?format=json&type=post_processed&force=false) for download csv.
3. after downloading csv  we would again create dataframe and then check the count of columns in dataframe 
4. count of columns in dataframe we would compare with same data of csv.
5. we would repeat this for both True/false value of config
I think this test-case is only possible check.

rohitpawar2811 committed:
Update superset/charts/post_processing.py 

Resolved pylint errors  by removing redefining name 'config' and third party import current_app placing at the top

rohitpawar2811 commented:
@betodealmeida 
I have updated the PR with suggested changes, could you please review it again?

betodealmeida commented:
> @betodealmeida I have updated the PR with suggested changes, could you please review it again?

Looks like some of the tests are failing, can you take a look at them?

rohitpawar2811 commented:
@mike-fischer1  Actually, I think this fix would not work, as post-processing logic is used in other flows and is breaking things like data-import for maintaining the sequence/order of data through appending indexes.

There are other ways we can try:

1. We can disable the post_processing while emailing the data, but then the feature of sending ordered data, as we saw in Explore, will break. [Link to the API](https://superset-dev.hotwax.io/api/v1/chart/628/data/?format=json&type=post_processed&force=false)

2. Alternatively, we could drop the index after the data is retrieved from the API in the Email-Flow by checking that we have to return CSV data."

reports/execute.py

    def _get_csv_data(self) -> bytes:
        url = self._get_url(result_format=ChartDataResultFormat.CSV)
        auth_cookies = machine_auth_provider_factory.instance.get_auth_cookies(
            self._get_user()
        )

        if self._report_schedule.chart.query_context is None:
            logger.warning("No query context found, taking a screenshot to generate it")
            self._update_query_context()

        try:
            logger.info("Getting chart from %s", url)
            csv_data = get_chart_csv_data(url, auth_cookies)
            **###Here we apply check for CSV_INDEX enable disable and if disable then drop the csv index.**
        except SoftTimeLimitExceeded as ex:
            raise ReportScheduleCsvTimeout() from ex
        except Exception as ex:
            raise ReportScheduleCsvFailedError(
                f"Failed generating csv {str(ex)}"
            ) from ex
        if not csv_data:
            raise ReportScheduleCsvFailedError()
        return csv_data

rohitpawar2811 committed:
Merge branch 'apache:master' into master

rohitpawar2811 committed:
Fix the CSV index column issue and make it configurable

rohitpawar2811 commented:
@betodealmeida 
I have updated my changes as previously the test-cases were failing, could you re-run the workflow tests on my latest changes

rohitpawar2811 commented:
> any updates on this?
@mike-fischer1 @Antonio-RiveroMartnez 
These changes would work

rohitpawar2811 commented:
@rusackas can you run checks on this PR?

rohitpawar2811 commented:
@betodealmeida could you please re-run the checks.

rohitpawar2811 committed:
Merge branch 'apache:master' into master

rohitpawar2811 committed:
Update execute.py: Fix the order of import statements

rohitpawar2811 committed:
Fix: long one liner and fixed the test case failure because in test_email gets the csv without index

rohitpawar2811 committed:
Fixed the test case failure by checking if csv contains index column then only apply removal of custom index column

rohitpawar2811 committed:
Merge branch 'apache:master' into master

rohitpawar2811 committed:
Fixed Pre-Commit check error

rohitpawar2811 committed:
Merge branch 'master' of https://github.com/rohitpawar2811/supersetMaster

romran commented:
Hello everyone :)
Any updates on this issue? We are using Version: 3.0.1 and getting index column in csv file when sending email report. 
Maybe someone know if there another fix? Or should we finish this PR?

rusackas commented:
> Or should we finish this PR?

I'm for it! @rohitpawar2811 are you able to follow up on this?

Commit:
Merge branch 'master' of https://github.com/apache/superset

Commit:
Improved: Make checks more better which can identify the index column

rusackas commented:
Enabling CI checks.... 🤞

rusackas commented:
@betodealmeida @Antonio-RiveroMartnez looks like CI passed 🎉 Any bandwidth to revisit the requested changes?

Antonio-RiveroMartnez commented on tests/integration_tests/fixtures/trends.csv:
Shouldn't this file have some content?

rohitpawar2811 replied:
Sure, here's the corrected version:

"Hi! Here, I made some changes in one commit and then undid it in the next commit. However, the file is still shown here, but in my repository, there are no changes in this file.

Should I simply make a reset commit and perform a force push so that this file won't appear here?"

Antonio-RiveroMartnez commented on superset/commands/report/execute.py:
```diff
@@ -254,6 +255,14 @@ def _get_csv_data(self) -> bytes:
+            if csv_data:
+                buf = BytesIO()
+                temp_df = pd.read_csv(StringIO(csv_data.decode("utf-8")))
+                if temp_df.columns[0].strip() == "Unnamed: 0":
```
Why `Unnamed: 0`? Can we move it out of the function and make it clear why is that the value we're comparing? Are we certain that no matter what the CVS file is, this is the value it gets with numbered columns?

rohitpawar2811 replied:
>> Are we certain that no matter what the CVS file is, this is the value it gets with numbered columns?

Yes I am sure about, and not able to find a better way to check that index column is present in csv or not

rohitpawar2811 replied:
okay got your point.

Antonio-RiveroMartnez reviewed (commented):
Hey @rohitpawar2811  Left a couple of questions. Thanks so much for taking care of this

rusackas commented:
I think @betodealmeida's change request is blocking it. It looks like CI will need to run again, but there seem to be a few unanswered questions on the thread. @SkinnyPigeon feel free to give it a code review if you'd like, that always helps. Mabye @Antonio-RiveroMartnez or @eschutho know where this stands in someone's/anyone's queue otherwise.

SkinnyPigeon commented:
Hey @rusackas, @rohitpawar2811, playing about with the API a bit, it seems that there is an Enum called [POST_PROCESSED](https://github.com/apache/superset/blob/5660f8e5542b78e098c42306633c182d9a631d63/superset/common/chart_data.py#L45) that is what is causing the additional index to be added. This is set [here](https://github.com/apache/superset/blob/5660f8e5542b78e098c42306633c182d9a631d63/superset/commands/report/execute.py#L162) in the report execution for the CSV datatype. If this is left blank, at least in the API, it returns the same data minus the index.

We could add the config check to this point instead. WDYT?

rohitpawar2811 commented:
> Hey @rusackas, @rohitpawar2811, playing about with the API a bit, it seems that there is an Enum called [POST_PROCESSED](https://github.com/apache/superset/blob/5660f8e5542b78e098c42306633c182d9a631d63/superset/common/chart_data.py#L45) that is what is causing the additional index to be added. This is set [here](https://github.com/apache/superset/blob/5660f8e5542b78e098c42306633c182d9a631d63/superset/commands/report/execute.py#L162) in the report execution for the CSV datatype. If this is left blank, at least in the API, it returns the same data minus the index.
> 
> We could add the config check to this point instead. WDYT?

@SkinnyPigeon 
Yes, initially when I looked into it, I found the same: that post-processing was adding the index. However, by making the check there, we were disturbing a lot of other checks. That's why I moved on from that approach.

Actually there are some ingesters that are using csv/data download api, when I make changes in Post_Processing that ingesters check failed.

I think we have to go with Post_Processing but it needs a lot of refectoring. WDYT?

SkinnyPigeon commented:
> > Hey @rusackas, @rohitpawar2811, playing about with the API a bit, it seems that there is an Enum called [POST_PROCESSED](https://github.com/apache/superset/blob/5660f8e5542b78e098c42306633c182d9a631d63/superset/common/chart_data.py#L45) that is what is causing the additional index to be added. This is set [here](https://github.com/apache/superset/blob/5660f8e5542b78e098c42306633c182d9a631d63/superset/commands/report/execute.py#L162) in the report execution for the CSV datatype. If this is left blank, at least in the API, it returns the same data minus the index.
> > We could add the config check to this point instead. WDYT?
> 
> @SkinnyPigeon I think this should work, but we need to check the test cases. Actually, in the past, I made changes inside to the post-processing logic, but I think your suggested check here makes more sense.
> 
> **Disturbing Part :**
> 
> * Making check on uri will remove csv index as well dataframe index.
> 
> Should we remove the DataFrame (text-email) index as well? What do you think?

With regards to where to make the change, I would suggest [here](https://github.com/apache/superset/blob/5660f8e5542b78e098c42306633c182d9a631d63/superset/commands/report/execute.py#L253). You could leave that line as is but then have a check for the config setting underneath and pass the URL to a function like this if we want the index removed:
``` python
def remove_post_processed(url: str) -> str:
    """Remove the type=post_processed parameter from the URL query string.
    
    Args:
        url (str): The URL to process.
    Returns:
        str: The URL with the type=post_processed parameter removed."""
    if "?" not in url:
        return url
    base_url, query_string = url.split("?", 1)
    params = query_string.split("&")
    filtered_params = [param for param in params if param != "type=post_processed"]
    filtered_query_string = "&".join(filtered_params)
    filtered_url = f"{base_url}?{filtered_query_string}"
    return filtered_url
```

This gives us something nice and easy to write some tests for:
```python
def test_remove_post_processed():
    url = "https://superset.com/?param1=value1&type=post_processed&param2=value2"
    expected = "https://superset.com/?param1=value1&param2=value2"
    assert remove_post_processed(url) == expected

def test_retain_other_parameters():
    url = "https://superset.com/?param1=value1&param2=value2"
    expected = "https://superset.com/?param1=value1&param2=value2"
    assert remove_post_processed(url) == expected

def test_no_post_processed_present():
    url = "https://superset.com/?param1=value1&param2=value2"
    expected = "https://superset.com/?param1=value1&param2=value2"
    assert remove_post_processed(url) == expected

def test_empty_query_string():
    url = "https://superset.com/?"
    expected = "https://superset.com/?"
    assert remove_post_processed(url) == expected

def test_no_query_string():
    url = "https://superset.com"
    expected = "https://superset.com"
    assert remove_post_processed(url) == expected
```

I'd leave the design decisions to the the code owners in terms of whether to remove it from the DataFrame as well, but I'd say the CSV is probably a good place to start for now
